Fix php7 igbinary serialization and unserialization

TysonAndre opened the pull request:
Fix php7 igbinary serialization and unserialization for references, arrays, and objects.

Fixes #1 and some occurences of #12

In php7, zvals are no longer stored as pointers (`zval*`) in objects and arrays (like in php5), but instead as values(`zval`).
In order to properly serialize references or shared values, use the underlying array/reference's reference count pointer. For objects, continue using object id. (like php5 does)

Fix the tracking of reference ids when serializing. Previously, the bugs in reference id generation caused the error "igbinary_unserialize_ref: invalid reference" when duplicate array values or array references existed. 
- The code was previously calling igbinary_serialize_array_ref twice for array references(One for the reference, another time for the array). This is different from php5(which does that only once), and this PR ensures it is called only once (so that unserialize works properly)
- Continue to call igbinary_serialize_array_ref only once for objects.
- Unserialize should add reference ids for same non-reference types it did in the php5 extension code(IS_STRING,IS_LONG,IS_NULL,IS_DOUBLE,IS_FALSE,IS_TRUE). It previously didn't(Did it for types not in that list), this fixes that.

Convert references' `zval`s to references immediately, so that when unserialized, the references will be shared

This includes https://github.com/igbinary/igbinary7/pull/8 - Fixes the segfault mentioned in https://bugs.php.net/bug.php?id=72134

Misc: 
- Refactor unserialize: Create function igsd_append_ref to add references when unserializing
- Fix some compiler warnings, use more suitable PHP7 macros in several places.

jakubsacha committed:
Add test for bug 72134

RahimovIR committed:
Fix memory leak / segfault

This test work https://bugs.php.net/bug.php?id=72134
https://github.com/igbinary/igbinary7/pull/7

TysonAndre committed:
Merge pull request #2 from RahimovIR/master

Fix memory leak / segfault (by RahimovIR)

TysonAndre committed:
Fix php7 igbinary serialization and unserialization

Fixes #1 and some occurences of #12

In php7, zvals are no longer stored as pointers (`zval*`) in objects and arrays (like in php5), but instead as values(`zval`).
In order to properly serialize references or shared values, use the underlying array/reference's reference count pointer. For objects, continue using object id. (like php5 does)

Fix the tracking of reference ids when serializing. Previously, the bugs in reference id generation caused the error "igbinary_unserialize_ref: invalid reference" when duplicate array values or array references existed. 

- The code was previously calling igbinary_serialize_array_ref twice for array references(One for the reference, another time for the array). This is different from php5(which does that only once), and this PR ensures it is called only once (so that unserialize works properly)
- Continue to call igbinary_serialize_array_ref only once for objects.
- Unserialize should add reference ids for same non-reference types it did in the php5 extension code(IS_STRING,IS_LONG,IS_NULL,IS_DOUBLE,IS_FALSE,IS_TRUE). It previously didn't(Did it for types not in that list), this fixes that.

Convert references' `zval`s to references immediately, so that when unserialized, the references will be shared

Fixes the segfault mentioned in https://bugs.php.net/bug.php?id=72134

This includes https://github.com/igbinary/igbinary7/pull/8 , which should be merged first

Misc: 

- Refactor unserialize: Create function igsd_append_ref to add references when unserializing
- Fix some compiler warnings, use more suitable PHP7 macros in several places.

EugenMayer commented:
Are the tests broken or the implementation?

TysonAndre-ContinuousIntegration commented:
@EugenMayer The tests are working properly. This PR reduces the number of test failures from 4 to 1.
This PR fixes many, but not all, bugs. The remaining test failure is cyclic recursive array unserialization.

There's one test failure remaining in https://github.com/igbinary/igbinary7/blob/master/tests/igbinary_009.phpt - var_dump() serializes one level of the cyclic recursive array references as `array(...)` instead of `&array(...)`
- I'm not sure yet if the test expectation is wrong (as of php7) or that is actually a bug remaining in the implementation. Checking `preg_replace('/&array/', 'array', $dump_act) === preg_replace('/&array/', 'array', $dump_act)` only for the `cyclic $a = array(&array(&$a))` would result in the test passing

On the master branch of igbinary/igbinary7, 4 of the tests are failing. https://travis-ci.org/igbinary/igbinary7/jobs/97431765

For this PR, only 1 test (Down from 4) is failing: 
https://travis-ci.org/igbinary/igbinary7/jobs/144047237 (Running with valgrind shows no leaks)
https://travis-ci.org/igbinary/igbinary7/jobs/144047231 (Same test run without valgrind)

EugenMayer commented:
very very interesting!

TysonAndre committed:
Create igbinary_019b.phpt testing cyclic arrays

The var_dump comparison in igbinary_019.phpt will be slightly weakened.
This test case makes sure that the deserialized arrays behave the exact same way the (value of the) original would.

TysonAndre committed:
Normalize value before comparing var_dumps of cyclic variable

Normalize by calling `serialize(unserialize($variable))` before calling igbinary_serialize. This allows us to check that var_dump is exactly the same.

igibnary_009b.phpt tests that the behavior of the original example (with a redundant reference) has not changed.

tricky commented:
Nice, hey @TysonAndre  could you join me on #igbinary on freenode?

uduncanu commented:
I'm getting segfaults with this change, and it seems to be linked to a change made in the reference handling code.  There's a fix in https://github.com/TysonAndre/igbinary7/pull/4 to make it logically the same as before but I think while still maintaining the improvements @TysonAndre has made.

TysonAndre commented:
@uduncanu do you have a test case or code snippet or project that reproduces the crash?
Or bin2hex of the serialized value?

How are you calling Igbinary(memcache, igbinary_serialize, etc?)?

TysonAndre committed:
Properly return references to values when unserializing

Before, this was modifying the **pointer** to the zval output variable,
which does nothing. (ZVAL_DEREF was after ZVAL_COPY)
Now, this modifies the **contents** of the zval output variable.

Handle references to an object differently from the values of that object
when serializing. See igbinary_046d.phpt in next commit.

TysonAndre committed:
Add tests of expected reference behavior

igbinary_046d.phpt is the only test which differs from what php5 does.
(Makes all/none of the uses of an object into references, depending on what the last ref was)

TysonAndre committed:
fixup tests/igbinary_bug72134.phpt

TysonAndre committed:
Fix a segfault when calling __wakeup() on a object that is a reference

Was passing $this as a reference to call_user_func* in some cases.
It should always be an object.
And fix cyclic reference serialization.
If the root is a reference, it should be unserialized as a reference.
Add tests of cyclic o

TysonAndre commented:
@uduncanu 
1. igbinary is able to serialize and unserialize references to objects, even cyclic references.
    In order to serialize complicated structures, this PR does the conversion to references as soon as possible.
2. I looked at my PR
   again and found a bug in the way I unserialized references, if there was a mix of references and non-references. I also started distinguishing between references and non-references (That bug should be fixed in  php5 as well.)
3. I found my PR was calling __wakeup with $this being a `zval` that was an IS_REFERENCE, should have passed it the IS_OBJECT that it was a reference to. Fixed and added a test
   https://github.com/igbinary/igbinary7/pull/13/commits/ef895e14f26d144fd584d70725c9d6bdb181158f
   https://github.com/igbinary/igbinary7/pull/13/commits/ce3c56d2fc7845dcf153e57a0f1c0aa61ffdea07

TysonAndre committed:
Add test of success case of references to Serializable

uduncanu commented:
@TysonAndre It generally looks better based on commit 2d2b195 although there were still issues, and I'll try everything up to and including 1e56913 tomorrow.  It looked like the remaining problems were to do specifically with references, and putting the data through unserialize(serialize($data)) before igbinary_serialize()ing  and unserializing seemed to stop PHP segfaulting.  That specific call is the only use case we have where we use the PHP function calls, with everything else usiing  the memcached library (which now appears to work, although we'll do further testing over the next few days).

Also, thanks for spending all this time looking into these problems! It genuinely is appreciated.
